### PR TITLE
Unset `PYTHONHASHSEED` for the hash-depedenet test

### PR DESCRIPTION
**Pull request by nzw0301**
## Motivation

Resolve https://github.com/optuna/optuna/issues/4028, which describes the motivation quite well.

## Description of the changes

Unset the environmental variable `PYTHONHASHSEED`. I've conformed `tox` doesn't fail anymore by adding the following lines to `tox.ini`

```
[testenv:pytest]
extras =
    test
    optional
deps =
    scikit-optimize
    cma
    botorch
commands = pytest tests/samplers_tests/test_samplers.py::test_reproducible_in_other_process
```

and run the following command on the CLI: 

```bash
tox -e pytest
```

### CODE REVIEW

**Comment by not522** [comment]
Could you restore the original seed at the end of `test_reproducible_in_other_process`?

**Comment by nzw0301** [comment]
@not522 Sure. I've updated the PR. 

Note that I replace `unsetenv` by following the recommendation of https://docs.python.org/3/library/os.html#os.unsetenv that reads

> Deletion of items in [os.environ](https://docs.python.org/3/library/os.html#os.environ) is automatically translated into a corresponding call to [unsetenv()](https://docs.python.org/3/library/os.html#os.unsetenv); however, calls to [unsetenv()](https://docs.python.org/3/library/os.html#os.unsetenv) don’t update [os.environ](https://docs.python.org/3/library/os.html#os.environ), so it is actually preferable to delete items of [os.environ](https://docs.python.org/3/library/os.html#os.environ).

**Review by not522 — APPROVED** [review]
Thank you! LGTM.

File: tests/samplers_tests/test_samplers.py [hunk 1]
```diff
@@ -1035,3 +1042,6 @@ def test_reproducible_in_other_process(sampler_class_index: int) -> None:
     assert not (hash_dict[0] == hash_dict[1] == hash_dict[2])
     # But the sequences are expected to be the same.
     assert sequence_dict[0] == sequence_dict[1] == sequence_dict[2]
+
+    if hash_seed is not None:
+        os.environ["PYTHONHASHSEED"] = hash_seed
```
Comment (not522) [inline tests/samplers_tests/test_samplers.py:1047]: Could you move these lines before asserts? If the assertion fails, `PYTHONHASHSEED` does not back to the original.

```suggestion
    if hash_seed is not None:
        os.environ["PYTHONHASHSEED"] = hash_seed

    assert not (hash_dict[0] == hash_dict[1] == hash_dict[2])
    # But the sequences are expected to be the same.
    assert sequence_dict[0] == sequence_dict[1] == sequence_dict[2]
```

**Review by not522 — CHANGES_REQUESTED** [review]

File: tests/samplers_tests/test_samplers.py (hunk 1 above)
Comment (nzw0301) [reply tests/samplers_tests/test_samplers.py]: Thank you for your comment and suggestions!

Comment (not522) [reply tests/samplers_tests/test_samplers.py]: The comment (This test should be~) doesn't match the code. Sorry, it is my mistake.

Comment (nzw0301) [reply tests/samplers_tests/test_samplers.py]: Thanks. I might misunderstand your comment, but I've slightly updated the comments not to specify `tox` case. What do you think?

Comment (not522) [reply tests/samplers_tests/test_samplers.py]: LGTM!

**Review by not522 — APPROVED** [review]

**Comment by toshihikoyanase** [comment]
I'm sorry for the delayed response. I did not investigate in detailed, but I'm a bit anxious about the failure of the test case might change `PYTHONHASHSEED` unexpectedly. Do you know if pytest has a mechanism like `try-except-finally` to ensure `PYTHONHASHSEED` can always be reverted?

**Comment by nzw0301** [comment]
Possibly https://docs.pytest.org/en/6.2.x/fixture.html#teardown-cleanup-aka-fixture-finalization can do that. I'll check it.

**Comment by nzw0301** [comment]
When I ran the target test: `test_reproducible_in_other_process`, and the following test, `test_another`, in https://github.com/nzw0301/optuna/blob/demo-unset-python-seed/tests/samplers_tests/test_samplers.py#L1038-L1069, where I commented out the other lines for this local test, I confirmed the hashseed value was restored in `test_another` even though an error was raised in the target function as @toshihikoyanase concerned.

```bash
PYTHONHASHSEED=1 pytest tests/samplers_tests/test_samplers.py


------------------------------------------------------------------------------------ Captured stdout call -------------------------------------------------------------------------------------
seed: None
....
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=================================================================================== short test summary info ===================================================================================
FAILED tests/samplers_tests/test_samplers.py::test_reproducible_in_other_process[0] - ValueError
FAILED tests/samplers_tests/test_samplers.py::test_reproducible_in_other_process[1] - ValueError
FAILED tests/samplers_tests/test_samplers.py::test_another - AssertionError: assert '1' is None
```

As we can see in the last line that comes from `test_another`, the hash seed is restored. I also note that the seed value was correctly unset in `test_reproducible_in_other_process`, which was confirmed by  `seed: None` line in the output.

File: tests/samplers_tests/test_samplers.py [hunk 2]
```diff
@@ -1016,8 +1018,29 @@ def objective(trial: Trial) -> float:
...
+    def restore_seed() -> None:
+        if hash_seed is not None:
+            os.environ["PYTHONHASHSEED"] = hash_seed
+
+    request.addfinalizer(restore_seed)
+
+
 @pytest.mark.parametrize("sampler_class_index", range(len(sampler_class_with_seed)))
-def test_reproducible_in_other_process(sampler_class_index: int) -> None:
+def test_reproducible_in_other_process(sampler_class_index: int, unset_seed_in_test: None) -> None:
```
Comment (toshihikoyanase) [inline tests/samplers_tests/test_samplers.py:1039]: I have two questions about type hint. `unset_seed_in_test` is not referred to in this test function, and I guess why `mypy` raise any error without any annotations. I know that it is a fixture and is called background.

And also, the type of `unset_seed_in_test` is set to `None`. Is it a return type of `unset_seed_in_test()`?

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your investigation and update. The fixture for finalization sounds great to me. I have questions about type hints.

File: tests/samplers_tests/test_samplers.py (hunk 2 above)
Comment (nzw0301) [reply tests/samplers_tests/test_samplers.py:1039]: Yes, your understanding is correct.

Comment (toshihikoyanase) [reply tests/samplers_tests/test_samplers.py:1039]: Thank you.

**Review by toshihikoyanase — APPROVED** [review]
I'm sorry for the delayed response. LGTM!

**Comment by toshihikoyanase** [comment]
@not522 I found that you self-requested a review. Please merge after your review.

**Review by not522 — APPROVED** [review]
LGTM

**Pull request merged**